crates/bevy_ecs/src/query/state.rs

```diff
@@ -40,6 +40,44 @@ impl<Q: WorldQuery, F: WorldQuery> FromWorld for QueryState<Q, F> {
     }
 }
 
+impl<Q: WorldQuery, F: WorldQuery> QueryState<Q, F> {
+    /// Converts this `QueryState` reference to a `QueryState` that does not access anything mutably.
+    pub fn as_readonly(&self) -> &QueryState<Q::ReadOnly, F::ReadOnly> {
+        // SAFETY: invariant on `WorldQuery` trait upholds that `Q::ReadOnly` and `F::ReadOnly`
+        // have a subset of the access, and match the exact same archetypes/tables as `Q`/`F` respectively.
+        unsafe { self.as_transmuted_state::<Q::ReadOnly, F::ReadOnly>() }
+    }
+
+    /// Converts this `QueryState` reference to a `QueryState` that does not return any data
+    /// which can be faster.
+    ///
+    /// This doesn't use `NopWorldQuery` as it loses filter functionality, for example
+    /// `NopWorldQuery<Changed<T>>` is functionally equivelent to `With<T>`.
+    pub fn as_nop(&self) -> &QueryState<NopWorldQuery<Q>, F> {
+        // SAFETY: `NopWorldQuery` doesn't have any accesses and defers to
+        // `Q` for table/archetype matching
+        unsafe { self.as_transmuted_state::<NopWorldQuery<Q>, F>() }
+    }
+
+    /// Converts this `QueryState` reference to any other `QueryState` with
+    /// the same `WorldQuery::State` associated types.
+    ///
+    /// Consider using `as_readonly` or `as_nop` instead which are safe functions.
+    ///
+    /// # SAFETY
+    ///
+    /// `NewQ` must have a subset of the access that `Q` does and match the exact same archetypes/tables
+    /// `NewF` must have a subset of the access that `F` does and match the exact same archetypes/tables
+    pub(crate) unsafe fn as_transmuted_state<
```

Comment:
Looks sound to me, but also more than a little bit cursed. Given that you are generally our "remove unsafe hacks at all costs" person, its interesting that you're advocating for this here when these are largely just incremental / superficial ergonomics wins. I love the way the code everywhere else looks now (and from my perspective I think it might be worth it), but I'd like to have one last discussion about how "worth it" this is.

* While this improves the "type legibility" of many of our apis, it also comes at the cost of some weird "non-safe-rust-semantics" behavior. The old apis, while more type-ey, also used safe rust semantics. Like if this PR represented the current state of main, and you opened a PR that switched to the "safer" QF approach, I wouldn't be surprised and I would be tempted to merge. People trying to understand whats going on here will likely be more than a little bit confused. We should weigh that, plus the "how cursed is bevy_ecs" meter going up (when ideally it should only be going down), against the clarity wins.
* This relies on us never adding a `Q::SomeOtherThing` field to `QueryState` to be valid. Unlikely to change any time soon, but also an easy mistake to make.
* Are we stuck with this forever or do we have a plan to remove it?

Reply:
I think we sort of want this kind of behaviour in some form or another to allow turning `Query<&mut T>` -> `Query<&T>` (which I think would be fairly valuable). I'll note that I specifically kept this method `pub(crate)` not `pub` like `as_readonly` and `as_nop` because
- the invariants seem fairly complex so not great to shove onto users :sweat_smile: 
- I feel like there probably is a safe API we can expose for "arbitrary" worldquery transformations 

The transmute is definitely,,, kind of sketchy :thinking:  I expect this could probably be done "safer" if we had the Q/F generics on QueryState be the state types rather than worldquery, and then moved all of the logic off of `QueryState`. It would then sort of be a no-op to turn `QueryState<Q::State>` -> `QueryState<Q::ReadOnly::State>` since the assoc types are the same. This would have the downside that it _isnt_ correct to use any arbitrary `Query<Q::State>` with a `Q` worldquery I think we'd lose a bit of type safety with this actually since it seems like a "good" thing to have some unsafe somewhere when you try to use a `QueryState` intended for a different `Q`.

I could imagine some potential future where we have
```rust
#[repr(transparent)]
struct QueryState<Q: WorldQuery>(RawQueryState<Q::State>);

fn transform_state_worldquery<Q: WorldQuery, NewQ: WorldQuery>(_: &QueryState<Q>) -> &QueryState<NewQ>
where
    Q: TransformableInto<NewQ> { ... }
```
Which would still be backed by a transmute but transmuting a repr transparent thing is relatively "easy" to soundness check and the repr(transparent) would forbid adding more fields. This seems a bit less cursed to me actually, even without the `fntransform_state_worldquery` this way of setting up the transmute soundness seems nice.

Reply:
Ok I'm starting to come around on this :)

> [On the topic of `QueryState<ActualState>`]. This would have the downside that it isnt correct to use any arbitrary `Query<Q::State>` with a Q worldquery

Yeah I think keeping QueryState paired to actual queries makes sense, given that the access encoded in it represents the state of a specific Query. We shouldn't make it easier to use mismatched QueryState accesses.

> [cart saying stuff about the old impl being safer]

Something I didn't address in my original comment is that allowing people to pass arbitrary fetches into "access" apis like iterators and get_unchecked_mut forces devs to enforce the "this state is appropriate to use with this fetch" rule at every call site. I do like that your impl centralizes these decisions.

> [cart saying stuff about not adding new associated type fields to QueryState]

Prior to merging this pr, we should add a big warning inside the QueryState struct along the lines of `// SAFETY NOTE: do not add any new associated type fields here from Q or F. Doing so might invalidate the QueryState::as_transmuted_state impl`.

There are "safer" ways to ensure these invariants (ex the QueryRef or QueryStateRef I outlined on discord), but they introduce more complexity than I want to take on, especially at this point in the release cycle.

So yeah add that comment and lets get this merged :)